pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/model/conditions/Failure.groovy

```diff
@@ -48,16 +48,7 @@ class Failure extends BuildCondition {
 
     @Override
     boolean meetsCondition(@Nonnull WorkflowRun r, Object context, Throwable error) {
-        Result execResult = getExecutionResult(r)
-        if (context instanceof Stage && execResult != Result.ABORTED && r.getResult() != Result.ABORTED) {
-            return error != null
-        }
-        Result errorResult = null
-        if (error != null) {
-            errorResult = Utils.getResultFromException(error)
-        }
-        return execResult != Result.ABORTED &&
-            (execResult == Result.FAILURE || r.getResult() == Result.FAILURE || errorResult == Result.FAILURE)
+        return combineResults(r, error) == Result.FAILURE
```

Comment:
A little scary since I don't completely understand what all of the edge cases were for previously, but I looked at the regression tests that were added for some of the changes and they still seem to be passing, so maybe now that the build result is not set until later a lot of the edge cases don't matter? Happy to add some more tests for this if there is anything in particular you think would be a problem.